src/utils.jl

```diff
@@ -31,13 +45,32 @@ julia> StructArrays._map_params(T -> Complex{T}, Tuple{Int32,Float64})
 (Complex{Int32}, Complex{Float64})
 ```
 """
-_map_params(f, ::Type{Tuple{}}) = ()
-function _map_params(f, ::Type{T}) where {T<:Tuple}
-    (f(tuple_type_head(T)), _map_params(f, tuple_type_tail(T))...)
-end
 _map_params(f::F, ::Type{NamedTuple{names, types}}) where {names, types, F} =
     NamedTuple{names}(_map_params(f, types))
 
+function _map_params(f::F, ::Type{T}) where {T<:Tuple, F}
+    if @generated
+        types = types_to_tuple(T)
+        ex = :()
+        for t ∈ types
+            push!(ex.args, :(f($t)))
+        end
+        ex
+    else
+        _map_params_recursive(f, T)
+    end
+end
+
+_map_params_recursive(f, ::Type{Tuple{}}) = ()
+function _map_params_recursive(f, ::Type{T}) where {T<:Tuple}
+    (f(tuple_type_head(T)), _map_params_recursive(f, tuple_type_tail(T))...)
+end
```

Comment:
```suggestion
_map_params_fallback(f, ::Type{T}) where {T<:Tuple} = map(f, fieldtypes(T))
```

I think the above should be enough (updating also other references).